### PR TITLE
Re sync fix

### PR DESCRIPTION
**Pull request by bigkevmcd**
## Problem

 
We are generating too many secrets for accessing the downstream cluster, because we are doing this concurrently.

## Solution

This adds a mutex per service-account, which will _reduce_ the number of secrets created, but not totally eliminate the problem.
 
### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:48]: I might be off-base here, but should we also include the "cluster id" as part of the lock? I'm thinking that in theory if this is called for the same user, this will lock for all clusters (since it's the same name and namespace), but in theory this wait wouldn't be useful since the create will only happen for one cluster.

File: pkg/serviceaccounttoken/secret_test.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret_test.go:368]: Do we need a reactor for leases here?

File: tests/v2/integration/serviceaccount/serviceaccounttoken_test.go
Comment (MbolotSuse) [inline tests/v2/integration/serviceaccount/serviceaccounttoken_test.go:1]: One comment - the integration tests here are mostly for tests that call out to Rancher's API. so this test, which is more of a package level integration tests, is a bit out of place here. It can stay here, just something to consider.

File: tests/v2/integration/serviceaccount/serviceaccounttoken_test.go
Comment (MbolotSuse) [inline tests/v2/integration/serviceaccount/serviceaccounttoken_test.go:75]: One other thing that we could check here is that the secret which is returned is the same secret by all 10 threads. Up to you though, I don't think that's strictly necessary.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Couple of small comments.

File: pkg/serviceaccounttoken/secret.go
Comment (jakefhyde) [inline pkg/serviceaccounttoken/secret.go:49]: Would it make sense to `TryLock` here and short circuit if it cannot be acquired? If a handler is already running for a given object, we won't have to do anything if it fails, as it will be re-enqueued anyway.

File: tests/v2/integration/serviceaccount/serviceaccounttoken_test.go
Comment (bigkevmcd) [reply tests/v2/integration/serviceaccount/serviceaccounttoken_test.go:1]: Which of the tests would you consider to be "integration tests"?

File: pkg/serviceaccounttoken/secret.go
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go]: I'm not sure how we'd determine the cluster here?

**Comment by moio** [comment]
@jakefhyde , @MbolotSuse: noob question, are you aware of the reasons why we use a generated name instead of a fixed name (depending on the service account)?

https://github.com/rancher/rancher/blob/cfcdd8816eb38f18995696299a98082edcee837d/pkg/serviceaccounttoken/secret.go#L119

From my naive perspective the whole problem here is we are generating too many secrets, where only one should exist (per serviceaccount). Concurrent requests should turn into no-op if the secret exists already.

Am I missing any relevant aspects?

File: pkg/serviceaccounttoken/secret.go
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:49]: The Go docs caution against TryLock

If it returns false we'd need to return an error to get a retry tho'?

File: pkg/serviceaccounttoken/secret.go
Comment (moio) [reply pkg/serviceaccounttoken/secret.go]: The ID would have to be passed to `EnsureSecretForServiceAccount` as a new parameter I guess, as the cluster is "encoded" inside of `secretsCache` and `clientSet` but it is not exposed (and those objects might not even be aware of the Rancher ID).

**Comment by MbolotSuse** [comment]
@moio answered offline.

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [reply pkg/serviceaccounttoken/secret.go]: We could do a new parameter or allow the context to contain an optional prefix that could be used as a prefix to the lock key. Either way, unless I'm missing something we do need something here or this could lock up concurrent requests to different clusters.

File: pkg/serviceaccounttoken/secret.go
Comment (jakefhyde) [reply pkg/serviceaccounttoken/secret.go:49]: I don't think you would need a retry:
- goroutine 1 acquires lock
- goroutine 2 attempts to acquire lock
at this point, either goroutine 1 exits succesfully, and goroutine 2 does not need to run on the object, or goroutine 1 returns an error, which would cause the object to be reenqueued anyway. By always waiting on the lock, you run the risk of resource exhaustion if thousands of goroutines end up waiting for locks. I'm worried with the current approach that any handler that needs the secret to exist for the service account is always locking. Is there something we can do to only lock when the secret doesn't exist i.e we need to create it? i.e. if the get/list fails, acquire the lock, attempt another get/list (in case of race conditions), and then create? This way we don't have to lock every time we attempt to get the object, but rather only on creation?

Comment (moio) [reply pkg/serviceaccounttoken/secret.go:49]: @bigkevmcd 

> If it returns false we'd need to return an error to get a retry tho'?

Correct. Any error returned by handlers ([apart from the special one](https://github.com/rancher/rancher/wiki/Tips-for-Writing-Rancher-Handlers#errors): `ErrIgnore`) will result in lasso requeuing the handler for later execution - [with a variable delay](https://github.com/rancher/rancher/wiki/Tips-for-Writing-Rancher-Handlers#enqueuing).

Comment (moio) [reply pkg/serviceaccounttoken/secret.go:59]: @jakefhyde this method is not called inside of handlers exclusively - it's called for instance very close to the `main` of the `cattle-cluster-agent`, or in `Serve` methods that publish APIs, or inside of `kontainer-engine`.

All this places, AFAIU, will behave differently if we return an error that was not returned before just because of wrong timing.

In principle we could set a parameter to tell if it should block or not, although I am not sure it is worth the trouble.

For now we will be sending our customer a version of this PR more or less as it currently is and first of all observe if and how much it helps, and also benchmark if and how much lock contention really happens. Then I guess we can revisit this discussion for general inclusion for all customers.

Does that make sense from your perspective?

Comment (jakefhyde) [reply pkg/serviceaccounttoken/secret.go:59]: It does but I still feel like we should only attempt to lock if the secret does not exist yet. For each handler there's only one point where they would need to create it, so I feel like it doesn't make any sense to lock if we aren't creating it.

Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:59]: I'll take another look at locking only-on-creation.

Comment (jakefhyde) [reply pkg/serviceaccounttoken/secret.go:59]: I think the initial issues we were seeing were due to the locking happening every update, so hopefully restricting to only on create will improve performance.

Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:59]: @jakefhyde I think those were related to the old Lease mechanism that was put in.

Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:59]: I think the problem with `lock-on-create` is that there's a race, we call `ServiceAccountSecret` which lists all secrets with a label matching the service account, then we iterate through that list, and check to see which one is "the one", and we attempt to delete the rest.

Then, if we didn't find a matching secret, we call `.Create()`.

While there's a small window here, I can easily reproduce the failure in tests, how close those are to the "real world" is debatable, but there's still a race-window.

**Comment by samjustus** [comment]
#44347

File: pkg/serviceaccounttoken/secret.go
Comment (jakefhyde) [reply pkg/serviceaccounttoken/secret.go:59]: Maybe we can separate that into locking if there no suitable secret OR if there is more than one secret? From my perspective, we will only need to delete multiple secrets if more than one exists, and create if there is not a matching secret, but during normal flow there should only ever be 1 secret unless the existing one is deleted or a second is created, but I think it makes sense not to lock all the time. 

I won't block on it since CI is passing but it would be nice to do some scale testing for provisioning a few clusters at the same time to see how the performance behaves.

**Review by jakefhyde — APPROVED** [review]

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:176]: I might be missing something, but couldn't this deadlock with the calling function `createServiceAccountSecret`? Since that function locks before it calls this function, the lock will be held by the calling function. This function will wait on the lock to proceed, which it will never get because the calling function is waiting for this one to finish.

I definitely might be missing something, let me know if I am wrong.

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:241]:
```suggestion
```

I think this is a duplicate of the above and can be removed.

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:51]: Nit: Might be good to specify what "LockPrefix" should be since there seems to be a convention for it (empty if in the same cluster, `utils.FormatPrefix(cluster.Name)` if it's not).

File: pkg/serviceaccounttoken/secret_test.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret_test.go:353]: Nit: Might be good to add a a test case for what happens with multiple secrets that already exist since the cleanup/create is part of the new locking mechanism.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Some nits, and a few more substantial comments on a few areas:
- I think there's a deadlock in the current code (could be wrong)
- There's a few places (in the `kontainer-engine` package) where we are creating a SA token on a downstream but pass an empty cluster name.

File: pkg/kontainer-engine/drivers/aks/aks_driver.go
Comment (bigkevmcd) [reply pkg/kontainer-engine/drivers/aks/aks_driver.go:1307]: It's not an easy call, we could decide to only lock if `lockKey != ""` which would only lock if we knew for sure we were creating for a Cluster.

But, there are some cases (like this) where we don't necessarily have a cluster identifier, despite the fact that we're creating for a cluster.

File: pkg/agent/cluster/cluster.go
Comment (bigkevmcd) [reply pkg/agent/cluster/cluster.go:81]: I added it to format "user" data for use, it doesn't feel like it's necessary here?

File: pkg/kontainer-engine/drivers/aks/aks_driver.go
Comment (MbolotSuse) [reply pkg/kontainer-engine/drivers/aks/aks_driver.go:1307]: I think this one does have a cluster identifier. The [cluster object](https://github.com/rancher/rancher/blob/4e2703bb4f157be02dbc2f2828d99a5b1c61bb73/pkg/kontainer-engine/cluster/cluster.go#L38) that [calls the driver](https://github.com/rancher/rancher/blob/4e2703bb4f157be02dbc2f2828d99a5b1c61bb73/pkg/kontainer-engine/cluster/cluster.go#L141) has a Name field. 

That being said, any option that you can use to pass the option down (such as changing the Info object or by adding a new parameter to the drivers) may have a wider effect. The only easy option would be using the context, which I think is a bit hacky.

We could try and skip this here, as long as we are aware of the performance impact on components that use this (essentially that the post-check operations can only be done for one cluster at a time). Provisioning team (@jakefhyde ) might be able to evaluate if that trade-off is ok.

File: pkg/serviceaccounttoken/secret.go
Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:176]: I had pondered this at the time...but here's my reasoning (and I've run the tests a lot, which while not perfect, were able to detect concurrency issues).

First call comes in...

```
EnsureSecretForServiceAccount
    serviceAccountSecret
        acquire lock (but only if there's more than one secret) ** this is really a lock around deletion **
            find secret ** and delete additional secrets **
        release lock
        return possible secret
    if serviceAccountSecret returned no secret
        createServiceAccountSecret
            use serviceAccountSecret to check for a secret
                (which will only lock if there's additional secrets, which should've been delete during the earlier lock)
        create the secret
    wait for the secret to be populated with the token
```

Second call comes in...

```
EnsureSecretForServiceAccount
    serviceAccountSecret
       _again_ acquire lock if there's more than one secret (which there could be because we're deleting them in the other Goroutine) but this will wait until the deletion lock is released
    if serviceAccountSecret returns no secret then as above, create the secret.
```

There is a race between using `serviceAccountSecret` to get a Secret (and getting `nil` back) and creating the secret, this is the tradeoff of reducing the scope of the lock.

Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:176]: To be clear, very happy to walk through this, my analysis could easily be wrong too 😄

File: pkg/kontainer-engine/drivers/aks/aks_driver.go
Comment (bigkevmcd) [reply pkg/kontainer-engine/drivers/aks/aks_driver.go:1307]: I agree that adding it to the context is hacky.

It doesn't seem wildly unreasonable to include the `Cluster.Name` in the `ClusterInfo` (it feels slightly surprising that it's not there, given the other fields) which would fix it.

Comment (MbolotSuse) [reply pkg/kontainer-engine/drivers/aks/aks_driver.go:1307]: I thought about that too - I was a bit hesitant on that one since the ClusterInfo type seems a bit unique (uses protobuf tags) and I didn't know if there was a reason for us to not touch that (i.e. if this is used in third-party drivers and could require them to be rebuilt).

That being said, the above is mostly me just not knowing the impact, if we can do it and it doesn't cause any regressions I think that's definitely the right answer.

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [reply pkg/serviceaccounttoken/secret.go:178]: I think the problem case comes in on this part:

```
if serviceAccountSecret returned no secret
        createServiceAccountSecret
            use serviceAccountSecret to check for a secret
                (which will only lock if there's additional secrets, which should've been delete during the earlier lock)
```

In this case, `createServiceAccountSecret` locks before calling `serviceAccountSecret`. I think that you are correct that `serviceAccountSecret` only locks if it finds more than one secret. However, I think that this is possible. Another rancher replica (or other third party application) could have created additional secrets in this time frame. I think that this would produce the deadlock. I think that you could replicate this in a test by having the reactor return 0 secrets on the first get call, but 2 secrets on the second get call.

File: pkg/serviceaccounttoken/secret.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret.go:223]: We don't seem to be checking this error - any way that we can check for an error and return early? Notably, I'm thinking that this could produce some extra secrets if the list call fails and we create a secret anyway (though this is unlikely in production).

File: pkg/serviceaccounttoken/secret_test.go
Comment (MbolotSuse) [inline pkg/serviceaccounttoken/secret_test.go:402]: Is this necessary? I would have thought that the code being tested should be adding this when it makes the create call (and that this should be something that we can look for/assert on).

**Review by MbolotSuse — COMMENTED** [review]
One comment on an un-handled error (and one on tests that doesn't necessarily need to be addressed), but outside of that LGTM on the basis that this won't fix all cases of this problem, but may improve the situation.

**Review by MbolotSuse — APPROVED** [review]
Approving given the following:
- This should make things better
- We know that this isn't a complete solution
- The impact of the issue can be substantial.

Please wait for a second approval and squash before/when merging.

File: pkg/kontainer-engine/drivers/aks/aks_driver.go
Comment (moio) [reply pkg/kontainer-engine/drivers/aks/aks_driver.go:1307]: So far in Bullseye's experience I have never heard of users creating very large amounts of clusters via Rancher at the same time - in a lot of cases that is done via Terraform or CAPI.

Even in that case, the slowdown we are adding here is minimal - seconds at most per Cluster - where the creation of the cluster itself normally takes minutes (from my experience with EKS and AKS at least).

Therefore I agree this particular optimization can be postponed.

File: pkg/capr/common.go
Comment (bigkevmcd) [reply pkg/capr/common.go:325]: `""` doesn't bypass locking completely?

https://github.com/rancher/rancher/pull/46312/files#diff-8a853a257f4c9e8e5812b706a4f6a56eaa32ff67e57f9ac5ac4f78f1e1659f04R59

It only bypasses it for the listing functionality if it's empty, but for `EnsureSecretForServiceAccount` it prefixes the mutex name, which has the effect (when provided) of narrowing the scope of the lock.

File: pkg/serviceaccounttoken/secret.go
Comment (moio) [inline pkg/serviceaccounttoken/secret.go:212]: AFAIU this never happens, IIUC, lockKey is always (expected to be) non-empty.

I'd suggest to remove the condition, as otherwise the lock would seem bypassable when it is not designed to ever be bypassed

File: pkg/capr/common.go
Comment (moio) [reply pkg/capr/common.go:325]: Right, I missed that.

I guess the original point still stands though - in this case we are sure we are talking about the local cluster, aren't we? If so we should pass "local" as the prefix, right?

**Review by moio — APPROVED** [review]
This can be merged, thanks for the long hard work here

**Pull request merged**
